lib/puma/runner.rb

```diff
@@ -88,6 +88,20 @@ def start_control
       @control = control
     end
 
+    def end_control(type)
+      if @control && [:halt, :stop].include?(type)
+        @control.send type, true
+        @control.binder.close
+        @control.binder.close_unix_paths
+      end
+    end
+
+    def close_control_io
```

Comment:
The control server itself should just handle this. I can't think of a circumstance where we want to stop the control server but want to keep its IOs open, right?

Reply:
i.e. no one should call close_control_io, everywhere should call end_control(type).

Reply:
> The control server itself should just handle this

The problem is that there isn't a control server class, and servers can also inherit binders.  So closing anything in the `Server` class is dicey.  Won't make that mistake again...

Re `close_control_io`, I think I can get rid of it, I'll need to add a `sync` arg to the `end_control` method.